chainbridge/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+        MODULE_ID.into_account()
+    }
+
+    fn chain_whitelisted(id: ChainId) -> bool {
+        return Self::chains(id) != None;
+    }
+
+    fn bump_nonce(id: ChainId) -> DepositNonce {
+        let nonce = Self::chains(id).unwrap_or_default() + 1;
+        <Chains>::insert(id, nonce);
+        nonce
+    }
+
+    /// Commits a vote in favour of the proposal and executes it if the vote threshold is met.
+    fn vote_for(
+        who: T::AccountId,
+        nonce: DepositNonce,
+        src_id: ChainId,
+        prop: Box<T::Proposal>,
+    ) -> DispatchResult {
+        // Use default in the case it doesn't already exist
+        let mut votes = <Votes<T>>::get(src_id, (nonce, prop.clone())).unwrap_or_default();
+
+        if !votes.votes_for.contains(&who) {
+            // Vote and store
+            votes.votes_for.push(who.clone());
+            <Votes<T>>::insert(src_id, (nonce, prop.clone()), votes.clone());
+
+            Self::deposit_event(RawEvent::VoteFor(src_id, nonce, who.clone()));
+
+            // Check if finalization is possible
+            if votes.votes_for.len() == <RelayerThreshold>::get() as usize {
+                Self::finalize_execution(src_id, nonce, prop)?
+            } else if votes.votes_for.len() > <RelayerThreshold>::get() as usize {
+                Err(Error::<T>::ProposalAlreadyComplete)?
+            }
+            Ok(())
+        } else {
+            Err(Error::<T>::RelayerAlreadyVoted)?
+        }
+    }
+
+    /// Commits a vote against the proposal and cancels it if more than (relayers.len() - threshold)
+    /// votes against exist.
+    fn vote_against(
+        who: T::AccountId,
+        nonce: DepositNonce,
+        src_id: ChainId,
+        prop: Box<T::Proposal>,
+    ) -> DispatchResult {
+        // Use default in the case it doesn't already exist
+        let mut votes = <Votes<T>>::get(src_id, (nonce, prop.clone())).unwrap_or_default();
+
+        if !votes.votes_against.contains(&who) {
+            // Vote and store
+            votes.votes_against.push(who.clone());
+            <Votes<T>>::insert(src_id, (nonce, prop.clone()), votes.clone());
+
+            Self::deposit_event(RawEvent::VoteAgainst(src_id, nonce, who.clone()));
+
+            // Check if cancellation is possible
+            if votes.votes_against.len()
+                > (<RelayerCount>::get() - <RelayerThreshold>::get()) as usize
+            {
+                Self::cancel_execution(src_id, nonce)?
+            }
+
+            Ok(())
+        } else {
+            Err(Error::<T>::RelayerAlreadyVoted)?
+        }
+    }
+
+    /// Execute the proposal and signals the result as an event
+    fn finalize_execution(
+        src_id: ChainId,
+        nonce: DepositNonce,
+        call: Box<T::Proposal>,
+    ) -> DispatchResult {
+        Self::deposit_event(RawEvent::ProposalApproved(src_id, nonce));
+        match call.dispatch(frame_system::RawOrigin::Signed(Self::account_id()).into()) {
```

Comment:
Yes that is intentional in that the validity is not something the bridge pallet deals with.

Perhaps you are referring to the resourceId mapping that was more recently added. This can be set by `root` to define what methods map to what resource IDs. The bridge uses it to resolve resource IDs to call names, so implicitly any call dispatched here will have already been "whitelisted" in that mapping. That of course only works in a trusted setup, in future we will have to verify on chain.

One problem I was trying to solve is how to identify calls uniquely. Lets say we want to whitelist `Pallet.method`, how would we specify that in the runtime? We can't rely on the call index because it may change. For the time being I've opted for the human readable name, but that won't work on-chain, wondering if you see any better options.

Reply:
That approach makes sense. Still, should we not have a check in the call that the resource is actually allowed/whitelisted? I don't see any checks against the resources.